Add guide how to install RKE2 cluster with SELinux enabled (#70)
suse-edge/suse-edge.github.io: docs/quickstart/rke2-selinux.md

---
sidebar_position: 7
title: RKE2 cluster with SELinux enabled
---

# Intro
SELinux, or Security-Enhanced Linux, is a security framework for Linux operating systems that provides an additional layer of access control and mandatory access controls (MAC) beyond the traditional discretionary access controls (DAC). Developed by the National Security Agency (NSA) in collaboration with the open-source community, SELinux aims to enforce fine-grained control over processes, files, and system resources to enhance the overall security of the system.

In this guide, RKE2 cluster will be installed with SELinux enabled.

## Prerequisites

> * 1 VM. Hint [SLE Micro on OSX on Apple Silicon (UTM)](https://suse-edge.github.io/docs/quickstart/slemicro-utm-aarch64) can be used.

Reviewer: maybe could be useful the info about minimum VM requirements to test it.

> * 1 VM. Hint [SLE Micro on OSX on Apple Silicon (UTM)](https://suse-edge.github.io/docs/quickstart/slemicro-utm-aarch64) can be used.

Reviewer: Only for UTM? Could be useful also the libvirt reference

As the `rke2-policy` was installed manually on the VM, some of its paths may not be created correctly, so the following commands will ensure that all the paths are fine.

SSH on the VM and run:
```shell
mkdir -p /var/lib/cni; \
mkdir -p /opt/cni; \
mkdir -p /var/lib/kubelet/pods; \
mkdir -p /var/lib/rancher/rke2/agent/containerd/io.containerd.snapshotter.v1.overlayfs/snapshots; \
mkdir -p /var/lib/rancher/rke2/data; \
mkdir -p /var/run/flannel; \
mkdir -p /var/run/k3s; \
restorecon -R -i /etc/systemd/system/rke2.service; \
restorecon -R -i /usr/lib/systemd/system/rke2.service; \
restorecon -R /var/lib/cni; \
restorecon -R /opt/cni; \
restorecon -R /var/lib/kubelet; \
restorecon -R /var/lib/rancher; \
restorecon -R /var/run/k3s; \
restorecon -R /var/run/flannel
```

It's time for the RKE2 cluster to be installed but before that it should be said that RKE2 must be running Selinux mode:

```shell
mkdir -p /etc/rancher/rke2 && echo "selinux: true" > /etc/rancher/rke2/config.yaml
```

Install RKE2 Using Install Script

```shell
curl -sfL https://get.rke2.io | INSTALL_RKE2_EXEC="server" RKE2_SELINUX=true INSTALL_RKE2_VERSION=v1.27.3+rke2r1 sh -

# Enable and Start RKE2

> systemctl enable rke2-server
> systemctl start rke2-server

Reviewer: ```suggestion
systemctl enable --now rke2-server.service
```

---
sidebar_position: 7
title: RKE2 cluster with SELinux enabled
---

# Intro
`SELinux`, or Security-Enhanced Linux, is a security framework for Linux operating systems that provides an additional layer of access control and mandatory access controls (`MAC`) beyond the traditional discretionary access controls (`DAC`). Developed by the National Security Agency (`NSA`) in collaboration with the open-source community, `SELinux` aims to enforce fine-grained control over processes, files, and system resources to enhance the overall security of the system.

> > In this guide, RKE2 cluster will be installed with SELinux enabled.

Reviewer 2: ```suggestion
> In this guide, we'll show you how to deploy an RKE2 cluster with SELinux enabled and enforcing.
```

## Prerequisites

> * 1 VM. Hint [SLE Micro on OSX on Apple Silicon (UTM)](https://suse-edge.github.io/docs/quickstart/slemicro-utm-aarch64) or [SLE Micro on X86_64 on libvirt (virt-install)](https://suse-edge.github.io/docs/quickstart/slemicro-virt-install-x86_64) can be used.

Reviewer 2: ```suggestion
* 1 VM. Hint [SLE Micro on OSX on Apple Silicon (UTM)](https://suse-edge.github.io/docs/quickstart/slemicro-utm-aarch64) or [SLE Micro on X86_64 on libvirt (virt-install)](https://suse-edge.github.io/docs/quickstart/slemicro-virt-install-x86_64) can be used as the base platform for validation here, but these instructions should work on any SLE Micro based system.
```

  - The VM should meet the [RKE2 requirements](https://docs.rke2.io/install/requirements#linuxwindows).

## Installation

### Installation for x86-64 architecture

> Once we've got the VM started and running, let's prepare the config to enable selinux mode in the RKE2 configuration file:

Reviewer 2: ```suggestion
Once we've got the VM started and running, let's prepare the config to enable SELinux mode in the RKE2 configuration file:
```

```shell

> mkdir -p /etc/rancher/rke2 && echo "selinux: true" > /etc/rancher/rke2/config.yaml

Reviewer 2: ```suggestion
mkdir -p /etc/rancher/rke2 && echo "selinux: true" >> /etc/rancher/rke2/config.yaml
```

```shell
curl -sfL https://get.rke2.io | INSTALL_RKE2_CHANNEL=stable INSTALL_RKE2_METHOD=rpm RKE2_SELINUX=true sh -

# Enable and Start RKE2
systemctl enable --now rke2-server.service
```

Now, the VM should be rebooted:

```shell
reboot
```

### Installation on arm64 architecture

As there are no RPM builds for RKE2 on `arm64` architecture, the `tarball` method will be used and the `rke2-selinux` policy will be installed manually.

#### Install rke2-selinux

The first thing that will be installed is the [rke2-selinux](https://github.com/rancher/rke2-selinux) policy.

Let's connect to the VM and run:

```shell
cat >> install-selinux.sh << 'END'
#!/bin/bash

# Install rpm-testing.rancher.io repository key and the rke2-selinux package
mkdir -p /var/lib/rpm-state
curl -o /root/public.key https://rpm-testing.rancher.io/public.key
curl -L -o /root/rke2-selinux.rpm https://github.com/rancher/rke2-selinux/releases/download/v0.15.testing.1/rke2-selinux-0.15-1.slemicro.noarch.rpm
rpmkeys --import /root/public.key

# Install RKE2 with SELinux
zypper install -y /root/rke2-selinux.rpm
END

chmod +x install-selinux.sh && transactional-update run /root/install-selinux.sh
```

> Now, the VM should be rebooted:

Reviewer 3: ```suggestion
Now, the VM should be rebooted for the transactional-update to finish properly:
```

```shell
reboot
```

After restarting the VM, we can verify that the policy was successfully installed as follows:

```shell
rpm -qa | grep rke2
```

#### Install RKE2

> As a second step, a RKE2 cluster will be installed, which will use the policy installed in the previous section.

Reviewer 2: ```suggestion
As a second step, an RKE2 cluster will be installed, which will use the policy installed in the previous section.
```

As the `rke2-policy` was installed manually on the VM, some of its paths may not be created correctly, so the following commands will ensure that all the paths are fine.

Let's connect to the VM and run:
```shell
mkdir -p /var/lib/cni
mkdir -p /opt/cni
mkdir -p /var/lib/kubelet/pods
mkdir -p /var/lib/rancher/rke2/agent/containerd/io.containerd.snapshotter.v1.overlayfs/snapshots
mkdir -p /var/lib/rancher/rke2/data
mkdir -p /var/run/flannel
mkdir -p /var/run/k3s
restorecon -R -i /etc/systemd/system/rke2.service
restorecon -R -i /usr/lib/systemd/system/rke2.service
restorecon -R /var/lib/cni
restorecon -R /opt/cni
restorecon -R /var/lib/kubelet
restorecon -R /var/lib/rancher
restorecon -R /var/run/k3s
restorecon -R /var/run/flannel
```

It's time for the RKE2 cluster to be installed but before that, RKE2 must be running Selinux mode:

```shell

> mkdir -p /etc/rancher/rke2 && echo "selinux: true" > /etc/rancher/rke2/config.yaml

Reviewer 2: Suggest ">>" just incase the user has already created `/etc/rancher/rke2/config.yaml` and put additional parameters in here already.

> mkdir -p /etc/rancher/rke2 && echo "selinux: true" > /etc/rancher/rke2/config.yaml

Reviewer 2: ```suggestion
mkdir -p /etc/rancher/rke2 && echo "selinux: true" >> /etc/rancher/rke2/config.yaml
```

> In this guide, RKE2 cluster will be installed with SELinux enabled.

## Prerequisites

* 1 VM. Hint [SLE Micro on OSX on Apple Silicon (UTM)](https://suse-edge.github.io/docs/quickstart/slemicro-utm-aarch64) or [SLE Micro on X86_64 on libvirt (virt-install)](https://suse-edge.github.io/docs/quickstart/slemicro-virt-install-x86_64) can be used.
  - The VM should meet the [RKE2 requirements](https://docs.rke2.io/install/requirements#linuxwindows).

## Installation

### Installation for x86-64 architecture

Once we've got the VM started and running, let's prepare the config to enable selinux mode in the RKE2 configuration file:

```shell
mkdir -p /etc/rancher/rke2 && echo "selinux: true" > /etc/rancher/rke2/config.yaml
```

Install RKE2 cluster

```shell
curl -sfL https://get.rke2.io | INSTALL_RKE2_CHANNEL=stable INSTALL_RKE2_METHOD=rpm RKE2_SELINUX=true sh -
```

> Now, the VM should be rebooted:

Reviewer 3: Is this really needed? Why?

Author: I think yes, because the command above installs multiple RPM packages

Reviewer 3: Then I think it should be a good idea to add the reason :)

```suggestion
Now, the VM should be rebooted for the transactional-update to finish properly:
```

Reviewer 2: If you're rebooting immediately after, do you need the `--now` flag on the `systemctl enable` call? Is there advantage to doing this prior to the reboot?

```shell
reboot
```

Let's connect to the VM and run:

```shell
# Enable and Start RKE2
systemctl enable --now rke2-server.service

> ```

Reviewer 3: If a reboot is needed for whatever reason, this step can be done before the reboot so after the host has been rebooted, RKE2 is already up.

> ### Installation for arm64 architecture

Reviewer 3: ```suggestion
### Installation on arm64 architecture
```

> As there are no RPM builds for RKE2 on `arm64` architecture, the `Tarball` method will be used and the `rke2-selinux` policy will be installed manually.

Reviewer 3: ```suggestion
As there are no RPM builds for RKE2 on `arm64` architecture, the `tarball` method will be used and the `rke2-selinux` policy will be installed manually.
```

#### Install rke2-selinux

The first thing that will be installed is the [rke2-selinux](https://github.com/rancher/rke2-selinux) policy.

Let's connect to the VM and run:

```shell
cat >> install-selinux.sh << 'END'
#!/bin/bash

> # Install SELinux Key and Package

Reviewer 3: ```suggestion
# Install rpm-testing.rancher.io repository key and the rke2-selinux package
```

Reviewer 3: Is there a non testing repository available? it looks really odd to install a package from rpm-testing 😓

mkdir -p /var/lib/rpm-state
curl -o /root/public.key https://rpm-testing.rancher.io/public.key
curl -L -o /root/rke2-selinux.rpm https://github.com/rancher/rke2-selinux/releases/download/v0.15.testing.1/rke2-selinux-0.15-1.slemicro.noarch.rpm
rpmkeys --import /root/public.key

# Install RKE2 with SELinux
zypper install -y /root/rke2-selinux.rpm
END

chmod +x install-selinux.sh && transactional-update run /root/install-selinux.sh
```

Now, the VM should be rebooted:

```shell
reboot

> ```

Reviewer 3: Is this really needed? Why?

Author: Again I think that when an RPM is installed on a snapshot based OS, a reboot is needed.
It even asks you for reboot after the installation

Reviewer 3: See my suggestion above

> In this guide, RKE2 cluster will be installed with SELinux enabled.

## Prerequisites

* 1 VM. Hint [SLE Micro on OSX on Apple Silicon (UTM)](https://suse-edge.github.io/docs/quickstart/slemicro-utm-aarch64) or [SLE Micro on X86_64 on libvirt (virt-install)](https://suse-edge.github.io/docs/quickstart/slemicro-virt-install-x86_64) can be used.
  - The VM should meet the [RKE2 requirements](https://docs.rke2.io/install/requirements#linuxwindows).

## Installation

### Installation for x86-64 architecture

Once we've got the VM started and running, let's prepare the config to enable selinux mode in the RKE2 configuration file:

```shell
mkdir -p /etc/rancher/rke2 && echo "selinux: true" > /etc/rancher/rke2/config.yaml
```

Install RKE2 cluster

```shell
curl -sfL https://get.rke2.io | INSTALL_RKE2_CHANNEL=stable INSTALL_RKE2_METHOD=rpm RKE2_SELINUX=true sh -
```

Now, the VM should be rebooted:

```shell
reboot
```

> SSH back on the VM and execute the following commands:

Reviewer: ```suggestion
Let's connect to the VM and run:
```

```shell
# Enable and Start RKE2

> systemctl enable rke2-server
> systemctl start rke2-server

Reviewer: ```suggestion
# Enable and Start RKE2
systemctl enable --now rke2-server.service
```
